docs/docs/adrs/adr-011-improving-test-confidence.md

```diff
@@ -0,0 +1,208 @@
+---
+sidebar_position: 12
+title: ADR Template
+---
+# ADR 11: Improving testing and increasing confidence
+
+## Changelog
+* 2023-08-11: Proposed, first draft of ADR.
+
+## Status
+
+Proposed
+
+## Context
+
+Testing, QA, and maintenance of interchain-security libraries is an ever-evolving area of software engineering we have to keep incrementally improving.
+
+The purpose of the QA process is to catch bugs as early as possible. In an ideal development workflow a bug should never reach production.
+
+A bug found in the specification stage is a lot cheaper to resolve than a bug discovered in production (or even in testnet).
+
+Ideally, all bugs should be found during the CI execution, and we hope that no bugs will ever even reach the testnet (although nothing can replace actual system stress test under load interacting with users).
+
+During development and testnet operation the most commonly found types of bugs are listed below.
+- improper iterator usage
+- unbounded array access/iteration
+- improper input handling and validation
+- improper cached context usage
+- non-determinism check (improper use of maps in go, relying on random values)
+
+Such bugs can be discovered earlier with better tooling. Some of these bugs can induce increases in block times, chain halts or introduce an attack surface which is difficult to remove if other systems have started depending on that behavior.
+
+#### Current state of testing
+Our testing suites consist of multiple parts, each with their own trade-offs and benefits with regards to code coverage, complexity and confidence they provide.
+
+## Unit testing
+Unit testing is employed mostly for testing single-module functionality. It is the first step in testing and often the most practical. While highly important, unit tests often **test a single piece of code** and don't test relationships between different moving parts, this makes them less valuable when dealing with multi-module interactions.
+
+Unit tests often employ mocks to abstract parts of the system that are not under test. Mocks are not equivalent to actual models and should not be treated as such.
+
+Out of all the approaches used, unit testing has the most tools available and the coverage can simply be displayed as % of code lines tested. Although this is a very nice and very easy to understand metric, it does not speak about the quality of the test coverage.
+
+Since distributed systems testing is a lot more involved, our reliance on simple unit testing tools should be minimized due to lack of useful feedback. Of course, sometimes unit tests are still necessary and helpful, but in some cases where unit tests are not helpful, we should use e2e or integration tests as appropriate rather than taking pains to add unit tests only to artificially increase coverage.
```

Comment:
This is mostly referring to code coverage results. You can have 100% code coverage and still not cover much of the actual code interactions going on.

Reply:
Changed to
```
Since distributed systems testing is a lot more involved, unit tests are oftentimes not sufficient to cover complex interactions. Unit tests are still necessary and helpful, but in cases where unit tests are not helpful e2e or integration tests should be favored.
```

Reply:
Yea seems reasonable, my comment was just trying to communicate that imo unit tests are always helpful, even in complex interactions, by testing individual components of the system. Higher level tests have their place as well for integrations